File: old/2022/julia/03.jl
# Advent of Code 2022, day 3
# (c) aichingert

module Day3

function getvalue(s,fst,scn)
	ch = '_'

	for c in s
		if c in fst && c in scn ch = c end
	end

	if isuppercase(ch) return 27 + Int(ch) - Int('A') 
	else return 1 + Int(ch) - Int('a') end
end

function part1(inp::Vector{String})
	ans = 0

	for s in inp
		half = length(s)÷2
		ans += getvalue(s,s[begin:half],s[half+1:end])
	end

	return ans
end

function part2(inp::Vector{String})
	ans = 0

	for i in 1:3:length(inp)
		ans += getvalue(inp[i],inp[i+1],inp[i+2])
	end

	return ans
end

inp = readlines("../input/03")

println("Part 1: $(part1(inp))")
println("Part 2: $(part2(inp))")

end

File: old/2015/rust/03.rs
// Advent of Code 2015, day 3
// (c) aichingert

use std::collections::HashSet;

fn solve(inp: &str, part: bool) -> usize {
    let mut s: usize = if part { 1 } else { 0 };
    let mut loc: Vec<(i32, i32)> = vec![(0, 0); s + 1];
    let mut p: HashSet<(i32, i32)> = HashSet::from([loc[s]]);

    inp.chars().for_each(|c| {
        update(&mut loc[s], &c);
        p.insert(loc[s]);
        if part {
            s = 1 - s;
        }
    });

    p.len()
}

fn update(loc: &mut (i32, i32), c: &char) {
    match c {
        '<' => loc.0 -= 1,
        '>' => loc.0 += 1,
        'v' => loc.1 -= 1,
        '^' => loc.1 += 1,
        _ => panic!("Invalid input"),
    };
}

fn main() {
    let input = std::fs::read_to_string("../input/03").unwrap();

    println!("Part 1: {}", solve(input.as_str(), false));
    println!("Part 2: {}", solve(input.as_str(), true));
}
